### PR TITLE
chore(configuration): Add @TenantValue for per-tenant value injection

### PR DESCRIPTION
**Pull request by fhuaulme**
**What is the problem this Pull Request is trying to solve?**

Spring applications can use `@Value` to get values from configuration. However, retrieving values based on current tenant is not allowed with `@Value`.
 
**What is the chosen solution to this problem?**

Introduce a new annotation `@TenantValue` that be injected to a field of a Spring component when instantiated.
A check is added to ensure the injected component is `prototype` scope (singleton scope should never be allowed to have `@TenantValue`).
 
**Please check if the Pull Request fulfills these requirements**
- [x] The PR commit message follows our [guidelines](../CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features, coverage should be over 75% in the new code)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] Related design / discussions / pages (not in JIRA), if any, are all linked or available in the Pull Request

### CODE REVIEW

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/10/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/10/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/11/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/11/)

File: daikon-spring/daikon-spring-configuration/src/main/java/org/talend/daikon/spring/configuration/TenantValue.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.talend.daikon.spring.configuration;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.FIELD })
+public @interface TenantValue {
```
Comment (sgandon) [inline daikon-spring/daikon-spring-configuration/src/main/java/org/talend/daikon/spring/configuration/TenantValue.java]: I think you need some javadoc here to explain the values pattern you are expecting here.

File: daikon-spring/daikon-spring-configuration/src/test/java/org/talend/daikon/spring/configuration/TenantValuePostProcessorTest.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        final ValidBean bean = new ValidBean();
+        Object beforeInit = processor.postProcessBeforeInitialization(bean, "validName");
+
+        // then
+        assertSame(beforeInit, bean);
+    }
+
+    private static class ValidBean {
+
+        @TenantValue("${value}")
```
Comment (sgandon) [inline daikon-spring/daikon-spring-configuration/src/test/java/org/talend/daikon/spring/configuration/TenantValuePostProcessorTest.java]: example using a a field name *value* can be misleading, could you use a less abstract name please ?

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/12/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/12/)

**Comment by fhuaulme** [comment]
> is it applicable today with the ConfigService or is it for possible config service change ?

No, ConfigService is accessed using a ConfigClient and there's no automatic `@Value` injection using Configuration Service.

> is it working with a hybrid Environment bean that can access local config and remote config alike ?

Yes it is, the property value retrieval is using the `Environment` Spring class that abstracts underlying property sources (remote or local).

> the postProcessAfterInitialization method is supposed to be called for each bean registered in spring. Isn't a big overhead at the start of the app ? is there any way to target the king of bean to reduce the application scope ?

Nope, there's only the BeanPostProcessor interface that performs this type of operation (thus for all scopes).

File: daikon-spring/daikon-spring-configuration/src/main/java/org/talend/daikon/spring/configuration/TenantValuePostProcessor.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        doWithFields(bean.getClass(), //
+                field -> {
+                    final TenantValue annotation = findAnnotation(field, TenantValue.class);
+                    final Object value = HELPER.replacePlaceholders(annotation.value(), placeholderName -> {
+                        final Tenant tenant = TenancyContextHolder.getContext().getTenant();
+                        final String tenantId = String.valueOf(tenant.getIdentity());
+                        final String configurationKey = tenantId + "." + placeholderName;
+
+                        LOGGER.debug("Looking for configuration key '{}' for tenant '{}'", placeholderName, tenantId);
+                        LOGGER.debug("Resolved tenant configuration key '{}'", configurationKey);
```
[the remark this answers is not in the document]
Comment (fhuaulme) [reply daikon-spring/daikon-spring-configuration/src/main/java/org/talend/daikon/spring/configuration/TenantValuePostProcessor.java]: 👍 , indeed.

File: daikon-spring/daikon-spring-configuration/src/main/java/org/talend/daikon/spring/configuration/TenantValue.java (hunk 1 above)
Comment (fhuaulme) [reply daikon-spring/daikon-spring-configuration/src/main/java/org/talend/daikon/spring/configuration/TenantValue.java]: 👍

File: daikon-spring/daikon-spring-configuration/src/test/java/org/talend/daikon/spring/configuration/TenantValuePostProcessorTest.java (hunk 2 above)
Comment (fhuaulme) [reply daikon-spring/daikon-spring-configuration/src/test/java/org/talend/daikon/spring/configuration/TenantValuePostProcessorTest.java]: 👍

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/13/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/13/)

File: daikon-spring/daikon-spring-configuration-etcd/src/main/java/org/talend/daikon/spring/etcd/EtcdPropertySource.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            final CompletableFuture<GetResponse> future = kvClient.get(ByteSequence.from(etcdKeyName, CHARSET));
+            final List<KeyValue> values = future.get().getKvs();
+            if (!values.isEmpty()) {
+                final KeyValue keyValue = values.get(0);
+                return keyValue.getValue().toString(CHARSET);
+            } else {
+                return null;
+            }
+        } catch (InterruptedException | ExecutionException e) {
+            e.printStackTrace();
```
Comment (sgandon) [inline daikon-spring/daikon-spring-configuration-etcd/src/main/java/org/talend/daikon/spring/etcd/EtcdPropertySource.java]: beurk

Comment (fhuaulme) [reply daikon-spring/daikon-spring-configuration-etcd/src/main/java/org/talend/daikon/spring/etcd/EtcdPropertySource.java]: Hehe. Of course 👍

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/14/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/14/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/40/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/40/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/52/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/52/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/53/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/53/)

**Pull request closed without merging**